test/pthread/test_pthread_proxying_cpp.cpp

```diff
@@ -68,7 +71,10 @@ void test_proxy_async() {
   // Proxy to returner.
   {
     queue.proxyAsync(returner.native_handle(), [&]() {
-      i = 3;
+      {
+        std::unique_lock<std::mutex> lock(mutex);
+        i = 3;
+      }
```

Comment:
Despite reading you description I've having trouble seeing how putting a lock around this one assignment fixes the issue.  Would making `i` and atomic also fix the issue?

Reply:
After reading https://stackoverflow.com/questions/17101922/do-i-have-to-acquire-lock-before-calling-condition-variable-notify-one I think i understand..  I had never known this about cond vars..

Reply:
No, the lock is important because of the condition variable. The condition variable needs to atomically check the condition and block if the condition is not met. The mechanism for combining those into an atomic action is the lock, so if the condition can change without the lock being held, the condition variable doesn't work as intended.